### PR TITLE
Improve errors location tracking

### LINKED ISSUE #14123: [Bug]: Some SyntaxErrors still ignore startLine option.
### 💻

- [x] Would you like to work on a fix?

### How are you using Babel?

Programmatic API (`babel.transform`, `babel.parse`)

### Input code

```js
require("@babel/parser").parse("import.meta", { sourceType:"script", startLine: 10 })
```

### Configuration file name

_No response_

### Configuration

N/A

### Current and expected behavior

Following up from https://github.com/babel/babel/issues/9015, it appears only some of the SyntaxErrors were fixed to take into account he `startLine` option. As you can see in the example above, it is still possible to get SyntaxErrors with the unadjusted line number. Others however do respect it.

### Environment

- Babel version: 7.16.7
- https://runkit.com/tolmasky/some-syntax-errors-still-ignore-the-startline-option

### Possible solution

_No response_

### Additional context

_No response_

### LINKED ISSUE #14144: [Bug]: expectContextual drops token information and throws generic "unexpected token" error.
### 💻

- [X] Would you like to work on a fix?

### How are you using Babel?

Programmatic API (`babel.transform`, `babel.parse`)

### Input code

I ran into this while fixing https://github.com/babel/babel/issues/14123, and the fix for that will include a fix for this too. If you try the following code:

```javascript
require("@babel/parser").parse("import * from 'pizza'", { sourceType: "module" });
```

See it here: https://runkit.com/tolmasky/expectcontextual-drops-token-information

You'll get `SyntaxError: Unexpected token (1:9)`, despite the code having enough information to say "Unexpected token, expect "as"", and in fact it appears like it was originally designed to do so. However, this information is lost between the conversion from `expectContextual` and `unexpected`.

### Configuration file name

_No response_

### Configuration

_No response_

### Current and expected behavior

Should say something like "Unexpected token, expect "as"",

### Environment

- Babel version: 7.16.8
- https://runkit.com/tolmasky/expectcontextual-drops-token-information

### Possible solution

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tolmasky**
### Summary

The main change in this PR is to remove our dependence on `getLineInfo` by instead relying more heavily on the `Position` versions of the various locations that are passed around during parsing. Aside from thus "naturally" accounting for `startLine` and `startColumn` with no additional work, this also avoids the pathalogical cases when `getLineInfo` can't rely on previously cached locations and must thus re-iterate through the entire file to re-generate line and column information, despite the fact that other parts of the program already have this information. Although this is perhaps not *too* terrible in the default `{ errorRecovery: false }` case of calling `parse`, it can be quite devastating to performance if you are running with `{ errorRecovery: true }`, like we do in RunKit.

### Approach

Perhaps the most important change is that I have added an `index` member to the `Position` class, which tracks the source index (the value represented by, e.g.,`State.[start|end]` vs. `State.[start|end]Loc`):

```typescript
class Position
{
  line: number;
  column: number;
  index: number;
}
```

The reason for this is that these two values *do* exist in a one-to-one mapping, and thus anywhere where they must be passed separately simply introduces the possibility of erroneously getting them out of sync. It is also almost always the case that various methods, from an interface perspective, either care about both, or just the `Position`, and thus we are not creating too many cases where we are creating extra objects during the parse. That is to say, the places where it is beneficial to *just* update `state.pos` all still behave that way since they are in tight loops, not updated and then calling into other functions (that we weren't already doing this for).

I have made this change fairly defensively by making index be a non-enumerable member of `Position`:

```javascript
Object.defineProperty(this, "index", { enumerable: false, value: index });
```

This way, I wouldn't have to re-generate *every* existing test to include this new addition, and it also won't show up unexpectedly if anyone is JSON-ing these trees or whatever. I am however happy to just make it a normal member variable though, if that is preferred. I just figured I'd start off doing it this way, as it is easy to change, and so that more attention can be paid to the fewer "meaningful" test changes instead of being bombarded with a PR that changes just about every test.

As such, a few key methods have changed:

#### raise()

```diff
- raise(pos: number, template: ErrorTemplate, ...params: any)
+ raise(template: ErrorTemplate, origin: Origin, ...params: any)
```

Where `Origin` is defined as:

```typescript
type Origin = {| node: Node |} | {| at: Position |};
```

Aside from the usage of `Position` instead of the previous `number` index, the idea here is to make the most common use cases super simple, as well as needing less Flow gymnastics (and IMO making it more readable). Namely, in the very common case where you want to `raise` an error "from" a node, the code looks something like this:

```javascript
this.raise(Errors.InvalidPropertyBindingPattern, {
  node: expr,
});
```

This then does the right thing by extracting the `Position` information from the node in question. Additionally, if we in the future want to change errors to support ranges (so that you can for example underline the entire node vs. just point to the beginning), we could now much more easily do this for all these calls, since we'd only need to update `raise` to also grab the `end` from the node too.

If however you still want to pass a specific location for the error, you use the `at` form:

```javascript
this.raise(Errors.UnexpectedImportExport, {
  at: this.state.startLoc,
});
```

This behaves much like the old version, except taking a `Position` instead of a `number` index. I believe both of these read closer to the `throw SomeError(info)` pattern they are approximating, as well as making it immediately obvious whether they are needing to do complex location stuff or just defaulting to the "standard" case of identifying a node as having an error.

#### unexpected()

The other big error change is to `unexpected`. `unexpected` used to take either an error template or a token, but now only does the token version:

```diff
- unexpected(pos?: ?number, messageOrType: ErrorTemplate | TokenType): empty
+ unexpected(loc?: ?Position, type?: ?TokenType): empty
```

The reason for this change is that you really don't "get" anything from using the `ErrorTemplate` form of `unexpected`, as it collapses down to just being a call to `raise` with `pos` set to `this.state.start` and automatically thrown. In other words, there's no special "unexpected messaging" or anything taking place, since the `ErrorTemplate` circumvents all that logic, and it's basically synonymous with just doing `throw this.raise()`. Despite this, it requires a fair amount of `flow` gymnastics to support these two unrelated types. As such, I replaced the few instances of this usage with `throw this.raise()` and left `unexpected()` to solely handle the token case.

#### expectContextual()

There's no big change in `expectContextual`, except that thanks to the above `unexpected` change, it can now throw a nicer error. Previously, something like `this.expectContextual(tt._as)` would throw an error message saying "Unexpected token", whereas now it will throw an error that says ""Unexpected token, expected as".

Closes #14144

### Side-effects

As a side-effect of this change, a number of function interfaces became simpler, now only requiring a location be passed in instead of the index/Position pair. We could actually do this with way more methods, but I limited it to solely those methods that were directly affected by my changes necessary for fixing errors here. This is particularly nice for methods that take optional locations, as this removes the ambiguity about whether you can pass *just* the integer index or if you must either pass both or neither.

Additionally, `getLineInfo` and `getLocationForPosition` have been removed, since they aren't needed anymore.

### More recoverable errors

I'd like to focus on this more after this PR, since errorRecovery is kind of what set us down this path, but just due to the nature of this change in particular, I was able to easily make `Errors.ElementAfterRest` a recoverable error, and the tests have been updated accordingly to reflect this.

Closes #14123.

| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #14123, Fixes #14144
| Patch: Bug Fix?          | Bug Fix
| Major: Breaking Change?  | No
| Minor: New Feature?      | No
| Tests Added + Pass?      | Yes
| Documentation PR Link    | No
| Any Dependency Changes?  | No
| License                  | MIT

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/14130"></a>

### CODE REVIEW

File: packages/babel-parser/src/parser/error.js
Comment (JLHwung) [inline packages/babel-parser/src/parser/error.js:81]: `this.state.startLoc` will always be updated when a new AST node is parsed. We can make `getLineInfo` a parser method and then read `startLine` from parser options.

File: packages/babel-parser/src/util/location.js
Comment (JLHwung) [inline packages/babel-parser/src/util/location.js:42]: if we support `startLine`, then we should also support `startColumn`.

**Comment by tolmasky** [comment]
I am looking at possibly just making the various `raise` methods take a Position instead of a... "pos" (it's really hard to talk about the difference between "loc" and "pos" when "loc" is a `Position`, however it appears in `location.js`, not `position.js`, and that in turn has a `SourceLocation`, which is different from both `pos` and `loc`. I wish "pos" was instead called "charIndex" or something...). Anyways, it appears that the actual places that rely on sending in custom "pos" are fairly constrained, and so far, all have access to the corresponding loc as well. This could potentially obviate the need for any of this instead of moving getLineInfo (a fairly inefficient function) around. However, there might be some tricky place where that isn't the case.

It would also be nice if "loc" and "pos" weren't separate entities that would need to be passed around together. For example, if `Position` simply contained a `charIndex` property, that way you could pass one "thing" around everywhere:

```javascript
export class Position {
  line: number;
  column: number;
  charIndex: number
}
```

But this is of course a breaking change and outside the scope of all this.

**Comment by tolmasky** [comment]
I have a significantly different change that accounts for column, etc. too. Will push either later today or tomorrow.

**Comment by tolmasky** [comment]
I'm going to try to get an updated PR submitted later today. Just writing a bunch of tests and stuff now.

**Comment by tolmasky** [comment]
OK, I have added a bunch of info to the main section above explaining this change. I am now in the process of trying something with the tests that would allow us to try every existing test with different startLine/startColumn changes and thus avoid having to figure out the right places to test this.

File: packages/babel-parser/src/util/location.js
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/util/location.js:20]: This should probably go in a minor, because it's a new observable property on `loc` objects in the AST (making it non-enumerable hides it from the fixtures, but it's still there :stuck_out_tongue:).

However, I would like to merge this PR as soon as possible because:
1. This PR also fixes different `errorRecovery` exceptions and some other bugs
2. This PR conflicts with virtually everything touching the parser, and rebasing it will not be easy.

For now, we can hide this index property in a new `const indexes = new WeakMap()` exported by this file, and do
```suggestion
    // this.index = index;
    indexes.set(this, index);
```

And replace `/([\w\.]+)\.index/g` with `indexes.get($1)`. Then, we can discuss if we want to expose it as a public property in Babel 7.17.0.

Comment (tolmasky) [reply packages/babel-parser/src/util/location.js:20]: Sounds good, can make that change.

File: packages/babel-parser/src/parser/error.js
Comment (tolmasky) [reply packages/babel-parser/src/parser/error.js:88]: I actually originally had this, but decided against it for a couple reasons:

1. I wanted it to read like throwing an error "throw SomeError(metadata)", which calls out what matters most at that line "hey, this is a MissingSemicolonError" or "hey, this is RestTrailingCommaError"  vs. front-loading the information that matters least, the associated metadata.  when reading the code, all the metadata, like doing throw "metadata on SomeError";
2. I think we will want potentially want to add more to this object later. As I mentioned in the above, we may eventually want to expand this to `{ at: length: }` for ranges (or `{ start, end }` or however you want to do that), and I think it would also eventually be nice to be able to supply context info (we currently have this info but it gets stringified for example, but can be very useful to the user to be able to do error.enumName vs. parsing message for enumName). As such, I wanted to put in the scaffolding for something that could carry additional info that could be easily read in the future. 
3. From my testing there seems to be zero performance benefit to avoiding object creation at this point, for two reasons. First, it is a seldom occurring event, as it happens once per error raise (vs. getLineInfo which was an iterative event that can lead to an O(N) slowdown per Error, where N = source length). Secondly, if we did consider it to be performance sensitive, then we'd for example be much better served by instead changing `raiseWithData` to take (missingPlugin?: Array<string>, code?: string)` instead of `{ missingPlugin?: Array<string>,  code?: string, }` (since this object `data` object gets thrown away immediately when yet another object is created on the corresponding call to `_raise`), and this change would basically affect *all* error raises "for free" transparently since it's an internal function call in `raise`. However, I also don't think this is a performance issue.

Comment (nicolo-ribaudo) [reply packages/babel-parser/src/parser/error.js:88]: Ok, sounds good :+1:

**Comment by tolmasky** [comment]
OK, I've gone ahead and implemented the `indexes` change, and removed the testing change to not complicate this PR further. My plan is to make 2 other PRs after this:

1. One that just reverts the last commit that switches `.indexes` to the WeakMap approach. We can use that PR to discuss whether want to use `.index` or not in the next minor patch (or sometime after).

2. I'll put the testing changes into their own separate PR, that way I can devote that one to explaining them and so forth.

File: packages/babel-parser/src/util/scope.js
Comment (JLHwung) [inline packages/babel-parser/src/util/scope.js:44]: Good catch!

**Review by JLHwung — COMMENTED** [review]
In this PR the `pos: Position` is expanded into an `origin: Origin`. Personally I still prefer the argument order of previous `.raise()` signature because the `params` are coupled with the error template.

```
raise(origin: Origin, template: ErrorTemplate, ...params: any)
```

**Comment by tolmasky** [comment]
So, just upfront, I know this is a somewhat stylistic change so if you are not convinced by the following arguments, then I will absolutely swap the order back. That being said, I just wanted to check that you read my first response to Nicolo regarding this [here](https://github.com/babel/babel/pull/14130#discussion_r785457863), and a couple additional thoughts:

1.  As I mentioned to Nicolo above, my hope is to eventually use the rider "options" argument to pass this (currently limited) error metadata to `raise` such that it can *also* be attached to the eventually produced `SyntaxError`. In an ideal world with associated types, it would be great to have something like `this.raise(Errors.InvalidEnumName, { node, enumName, suggestion })`, such that someone writing dev tools would have access to these important items without having to parse error messages. Without fancy types though, even just `this.raise(Errors.InvalidEnumName, { at: location, metadata })` feels more akin to how this would work in a pure-throw-no-recovery code-style of doing `throw Errors.InvalidEnumName(location, metadata)`, with the added benefit of "named parameters". I actually started doing this when I was working on this change, which is where that swap originally happened, and then when I scaled the PR back, I decided to keep the swap just to float the above thinking.  All this to say, I am onboard with params mattering, and would like to make it even more readable in a future commit, by "having our cake and eating it too" and making all associated data feel attached to the template. I figured I'd do this change now as a very small number of Errors actually currently take advantage of params (~14%), and I'd love to increase that going forward.

2. As I was making my way though this, it became clear that like `unexpected`, the default situation is to throw the error from "here", and thus in a later change we could probably further simplify a lot of this code down to `this.raise(Errors.Whatever)` and infer the right place from lacking Origin argument, IOW, the function interface could be changed to:

    ```javascript
    raise(error: ErrorTemplate, origin?: Origin, ...params);
    ```

    Thus, the common case would be to just do `this.raise(Errors.Whatever)`, and we'd know what to do. This would compare to the (IMO) more confusing option of: `this.raise(null /*what is this?*/, Errors.Whatever);` (which we currently do with `unexpected(null, token)`). It happens to be that the default (currently) for `unexpected` is also to not pass the expected token, and thus this is "felt" less since you are much more often just taking the default for both: `this.unexpected()`. If we get better at reporting what we'd like to see however (like for example I know a few places where it would be nice to have something like `this.unexpected(null, Identifier)` (which you can't currently do since there's no single 'identifier' token, but you can imagine a Token | Concept which would allow that), you'll see more and more `this.unexpected(null, thing)`.

3. This last point kind of segue-ways to the most "style-y" argument that really just comes down to my feelings as someone expecting to soon wade through all the errors to try to make as many of them recoverable as possible ;), which is that the thing I want to stand out is what the error is, and not where the error happens to be coming from (which is always some variation of "here", in the boring case of `this.state.startLoc` or `node`, it becomes fairly monotonous noise, and in the more complicated cases of "position mechanics", very much stands in the way of just figuring out what the intent of the error is, as it is more likely to trigger a line wrap and thus for the Error "name" to be 2 lines removed from the error `raise`). 

Anyways, as I stated earlier, please don't let the wall of text make you feel that am unwilling to swap it back , just wanted to share various "error thoughts" I had while making my way through this, and I am happy to swap it back if they are not convincing.

File: packages/babel-parser/src/parser/util.js
Comment (tolmasky) [reply packages/babel-parser/src/parser/util.js:288]: So, this was difficult to understand, but *I think* this is what's going on:

1. The intent, as I understand it, of `refExpressionErrors` is to "hold onto" things that would be errors in expressions, but *not* in patterns.
2. The only places patterns can appear is in variable declarations, function parameters, and assignment expressions (which are ambiguous until we reach an `=` sign).
3. In the first two, `refExpressionErrors` is just passed down as `null`, so it is irrelevant for our discussion here.
4. However, if it the current expression was determined to be an assignment expression in `parseMaybeAssign`, then `refExpressionErrors` will have already been cleared since we know this *was* a pattern ( https://github.com/babel/babel/blob/5b4d55d6e532e3c2cd7284fe951fc792108a62af/packages/babel-parser/src/parser/expression.js#L323 )
5. As such, I *think* the check in `parseUpdate` never passes, as by this point `refExpressionErrors` has been cleared. `parseUpdate` is the only place that passes `false` for `andThrow` (and also the only place that bothers checking the result).

This is why no tests fail. I've done my best to construct expressions that find some other path where this line would have `refExpressionErrors` and rely on andThrow being false to avoid throwing them, but I haven't been able to. If you can confirm the above, I can try to modify the code further to reflect how it's being used.

Comment (tolmasky) [reply packages/babel-parser/src/parser/util.js:288]: Just want to add that I think it's totally possible that there is a relevant case that's handled by this and simply not tested, and I'm happy to change this code back to that, and also add said test if we can find it.

Comment (tolmasky) [reply packages/babel-parser/src/parser/util.js]: OK I went ahead and restored the old behavior I think, although I still can't seem to figure out how it behaves differently, but figured it wasn't worth blocking this PR on figuring that out. I can file a bug on exploring this further. We should try to get a test in there or figure out why it's never entered.

**Review by JLHwung — APPROVED** [review]

**Review by nicolo-ribaudo — APPROVED** [review]
Thanks! I noticed that there are a bunch of `FIXME:` comments, but they mostly seem about "explain why it's safe to get the prev/next pos here".

File: packages/babel-parser/src/util/location.js
Comment (pieh) [reply packages/babel-parser/src/util/location.js]: @nicolo-ribaudo @tolmasky despite swap to `WeakMap` just having typed member in https://github.com/babel/babel/blob/b05dad7fed07672514fa6d0d21ce4c1e2c3a6f79/packages/babel-parser/src/util/location.js#L13

seems to actually init this field with `void 0` ( see https://www.runpkg.com/?@babel/parser@7.16.10/lib/index.js#9 ).

This did trip some of our tests ( TBF the test was in practice to strict and didn't allow for any uncounted for fields, so we will be relaxing them a bit ), but most importantly, I just want to check with you wether this is acceptable for you given the comments above

Comment (tolmasky) [reply packages/babel-parser/src/util/location.js]: Hi @pieh, I actually reported this immediately after we shipped as I also noticed it: https://github.com/babel/babel/issues/14182. Apologies though, as I had the discussion with @nicolo-ribaudo offline, where he advised that it was probably better to just hash it out here: https://github.com/babel/babel/pull/14174, as we might just be adding them in fully in the next minor update. But I'll let @nicolo-ribaudo comment more here as I am definitely open to removing that line for the next patch update if its causing any issues.

Comment (pieh) [reply packages/babel-parser/src/util/location.js]: There is no problem here from my side (as I mentioned - our tests/validation were just too strict and brittle because of that - I really should just validate keys we do need (which is `line`/`column`, but don't enforce that they are only possible fields in `Position`). This is especially true as `babel` is not the only "source" we are handling errors from. 

My comment was just to make sure you are aware as this sounds like something you wanted to avoid in patch release. For us it really doesn't matter (in practice, and not semantics) if it's a patch or minor given that right now we do use `^` version selector (and not pinned version or `~` for pinned minor)

Comment (tolmasky) [reply packages/babel-parser/src/util/location.js]: Sure thing, and again, apologies! This one's on me, I just forgot to delete the declaration after switching the uses of it.

**Pull request merged**
